linkerd/timeout/src/lib.rs

```diff
@@ -1,137 +1,141 @@
+
+pub use self::failfast::{FailFast, FailFastError, FailFastLayer};
+pub use self::idle::{Idle, IdleError, IdleLayer};
+pub use self::probe_ready::{ProbeReady, ProbeReadyLayer};
+
+/// A timeout that wraps an underlying operation.
+#[derive(Debug, Clone)]
+pub struct Timeout<T> {
+    inner: T,
+    duration: Option<Duration>,
+}
+
+#[pin_project]
+pub enum TimeoutFuture<F> {
+    Passthru(#[pin] F),
+    Timeout(#[pin] time::Timeout<F>, Duration),
+}
+
+//===== impl Timeout =====
+
+impl<T> Timeout<T> {
+    /// Construct a new `Timeout` wrapping `inner`.
+    pub fn new(inner: T, duration: Duration) -> Self {
+        Timeout {
+            inner,
+            duration: Some(duration),
+        }
+    }
+
+    pub fn passthru(inner: T) -> Self {
+        Timeout {
+            inner,
+            duration: None,
+        }
+    }
+}
+
+impl<P, S, Req> Proxy<Req, S> for Timeout<P>
+where
+    P: Proxy<Req, S>,
+    S: tower::Service<P::Request>,
+{
+    type Request = P::Request;
+    type Response = P::Response;
+    type Error = Error;
+    type Future = TimeoutFuture<P::Future>;
+
+    fn proxy(&self, svc: &mut S, req: Req) -> Self::Future {
+        let inner = self.inner.proxy(svc, req);
+        match self.duration {
+            None => TimeoutFuture::Passthru(inner),
+            Some(t) => TimeoutFuture::Timeout(time::timeout(t, inner), t),
+        }
+    }
+}
+
+impl<S, Req> tower::Service<Req> for Timeout<S>
+where
+    S: tower::Service<Req>,
+    S::Error: Into<Error>,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = TimeoutFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Req) -> Self::Future {
+        let inner = self.inner.call(req);
+        match self.duration {
+            None => TimeoutFuture::Passthru(inner),
+            Some(t) => TimeoutFuture::Timeout(time::timeout(t, inner), t),
+        }
+    }
+}
+
+impl<C> Connect for Timeout<C>
+where
+    C: Connect,
+    C::Error: Into<Error>,
+{
+    type Connected = C::Connected;
+    type Error = Error;
+    type Future = Compat<TimeoutFuture<Compat01As03<C::Future>>>;
+
+    fn connect(&self) -> Self::Future {
+        let inner = self.inner.connect();
+        match self.duration {
+            None => TimeoutFuture::Passthru(inner.compat()),
+            Some(t) => TimeoutFuture::Timeout(time::timeout(t, inner.compat()), t),
+        }
+        .compat()
+    }
+}
+
+impl<F, T, E> Future for TimeoutFuture<F>
+where
+    F: Future<Output = Result<T, E>>,
+    E: Into<Error>,
+{
+    type Output = Result<T, Error>;
+    #[project]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        #[project]
+        match self.project() {
+            TimeoutFuture::Passthru(f) => f.poll(cx).map_err(Into::into),
+            TimeoutFuture::Timeout(f, duration) => {
```

Comment:
We could consider using select here now too. I don't think we need to change it as part of this PR, but it may be worth considering just selecting over the call on `inner` and a `delay_for(duration)`. If there is no change in performance it may at least simplify the implementation a little.

Reply:
I didn't do that here because if we replaced this with an async function using `select!` we would need to add a layer of boxing in order to return it from a `Service` impl. 

In a follow up after we're on `std::future` on master, I want to look into switching to nightly Rust to allow returning `impl Future` as `Service`'s `Future` type, and rewrite a lot of our manual future implementations with async-await.

Reply:
Sounds good